feat: s3 plugin

dtbuchholz opened the pull request:
## Summary

Adds an S3 plugin with upload/download functionality. 

## Details

See the README and `example.ts` for more info. The basic pattern is to configure required params, like access keys and a predefined bucket name, and then you can prompt for writing a local file to the bucket, or downloading it locally:
```ts
const s3Plugin = new S3Plugin({
  credentials: {
    accessKeyId: "AKEXAMPLES3S",
    secretAccessKey: "SKEXAMPLES3S,
  },
  region: "us-east-1",
  bucket: "virtuals-game-bucket",
  endpoint: "http://localhost:8014",
  sslEnabled: false,
  forcePathStyle: true,
});

// initialize and start agent...

const task1 = "Upload a file to the S3 bucket at the path `./package.json`";
const task2 =
  "Upload a file to the S3 bucket at the path `./README.md` with key `hello/world` and use a signed URL with a TTL of 900 seconds";
const task3 =
  "Download a file from the S3 bucket at key `hello/world` to the a file at `./test.txt`";

```
This will log something like:
```
-----[S3 Bot]-----
Environment State: {"accessKeyId":"AKEXAMPLES3S","secretAccessKey":"SKEXAMPLES3S","region":"us-east-1","bucket":"virtuals-game-bucket","endpoint":"http://localhost:8014","sslEnabled":"false","forcePathStyle":"true"}

-----[S3 Bot]-----
Action State: ...

-----[S3 Bot]-----
Performing function upload_file with args {"file_path":{"value":"./package.json"},"object_key":{"value":null},"use_signed_url":{"value":false},"ttl":{"value":3600}}.

-----[S3 Bot]-----
Uploading file './package.json' at key '1738106514127-package.json'

-----[S3 Bot]-----
Function status: done - File uploaded:
{"success":true,"url":"http://localhost:8014/virtuals-game-bucket/1738106514127-package.json"}.

-----[S3 Bot]-----
Performing function upload_file with args {"file_path":{"value":"./README.md"},"object_key":{"value":"hello/world"},"use_signed_url":{"value":true},"ttl":{"value":900}}.

-----[S3 Bot]-----
Uploading file './README.md' at key 'hello/world'

-----[S3 Bot]-----
Function status: done - File uploaded:
{"success":true,"url":"http://localhost:8014/virtuals-game-bucket/%241738106514127-package.json?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=AKEXAMPLES3S%2F20250128%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250128T232156Z&X-Amz-Expires=3600&X-Amz-Signature=7b984b440b1c22080df97c79b02140b49c39467096e81e87841669f38f2079e0&X-Amz-SignedHeaders=host&x-amz-checksum-mode=ENABLED&x-id=GetObject"}.

-----[S3 Bot]-----
Performing function download_file with args {"object_key":{"value":"hello/world"},"file_path":{"value":"./test.txt"}}.

-----[S3 Bot]-----
Downloading object at key 'hello/world'

-----[S3 Bot]-----
Function status: done - File downloaded to: ./test.txt.
```

julianapeace commented:
Thanks for your contribution! Tested it and it is working - just some small feedback

The .env.example is missing the value GAME_API_KEY

And this is a recurring behavior

> -----[S3 Bot]-----
Performing function upload_file with args {"file_path":{"value":"/path/to/local/file.txt"},"object_key":{"value":"my_file.txt"},"use_signed_url":{"value":"false"},"ttl":{"value":"3600"}}.

> -----[S3 Bot]-----
Function status [failed]: File path is required and must exist.

bot seems to realize it is a mistake and intends to correct it but continues to exhibit the same behavior repeatedly. Not sure where it's getting the file_path either.

dtbuchholz commented:
@julianapeace thanks for the review! i updated the `.env.example` and slightly altered the example so that it'll just run the tasks and exit, which eliminates the weird behavior. some additional thoughts below.

> The .env.example is missing the value GAME_API_KEY

in original logic (before the change I just pushed), I kept the `.env.example` file to only the s3 relevant variables because the `GAME_API_KEY` a prerequisite for using the framework as a whole, so it feels a bit redundant to add it? alternatively, I could create a root `.env.example` that includes _all_ of the possible env vars across all plugins (this is what Eliza does). the only other plugin that has an env file is the `onChainActionsPlugin` (which doesn't include the API key btw).

> bot seems to realize it is a mistake and intends to correct it but continues to exhibit the same behavior repeatedly. Not sure where it's getting the file_path either.

i was hoping you'd have more clarity on why that happens! the plugin is just an available action the agent can take, and it simply exposes upload/download functionality. i don't know enough about the GAME runtime loop to know exactly why it's executing an action when it shouldn't. but, instead of running the `await agent.step(...) ...` logic in the example, I removed that so the bot simply executes the tasks, and then the program exits. this way, we can at least get the PR merged, but I don't think the erroneous agent behavior is directly related to the s3 plugin?

julianapeace reviewed (approved):
Looks great! Thank you for your contribution!
